src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "alt_hold.h"
+
+#ifdef USE_ALTHOLD_MODE
+
+#include "drivers/time.h"
+#include "flight/failsafe.h"
+#include "flight/imu.h"
+#include "flight/position.h"
+#include "sensors/acceleration.h"
+#include "sensors/barometer.h"
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "osd/osd.h"
+#include "common/printf.h"
+#include "common/maths.h"
+#include "math.h"
+#include "build/debug.h"
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 3);
+
+PG_RESET_TEMPLATE(altholdConfig_t, altholdConfig,
+    .velPidP = 30,
+    .velPidD = 0,
+
+    .altPidP = 75,
+    .altPidI = 20,
+
+    .minThrottle = 6,
+    .maxThrottle = 65,
+);
+
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+static float getCurrentAltitude(altHoldState_s* altHoldState)
+{
+#ifdef USE_BARO
+    if (sensors(SENSOR_BARO) && baroIsCalibrated()) {
+        return 0.01f * baro.altitude;
+    }
+#endif
+    float rawAltitude = 0.01f * getEstimatedAltitudeCm();
+    if (ABS(altHoldState->smoothedAltitude) < 0.01f) {
```

Comment:
This is ugly. Why 1cm around ground? Comparing to 0 may be better (and float errors should not cause problem here)

Also, this may cause jumps when hovering at 0m ( = close to ground). About the wort time for it to happen